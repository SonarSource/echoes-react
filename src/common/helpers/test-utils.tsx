/*
 * Echoes React
 * Copyright (C) 2023-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { RenderOptions, render as rtlRender } from '@testing-library/react';
import userEvent, { Options as UserEventsOptions } from '@testing-library/user-event';
import React, { ComponentProps, PropsWithChildren } from 'react';
import { IntlProvider } from 'react-intl';
import { PropsWithLabels } from '~types/utils';
import { TooltipProvider } from '../../components/tooltip';

export function render(
  ui: React.ReactElement,
  options?: RenderOptions,
  userEventOptions?: UserEventsOptions,
) {
  return {
    ...rtlRender(ui, { wrapper: ContextWrapper, ...options }),
    user: userEvent.setup(userEventOptions),
  };
}

function ContextWrapper({ children }: PropsWithChildren<{}>) {
  return (
    <IntlProvider defaultLocale="en-us" locale="en-us">
      <TooltipProvider delayDuration={0}>{children}</TooltipProvider>
    </IntlProvider>
  );
}

export type OmitPropsWithLabels<T extends React.JSXElementConstructor<any>> = Partial<
  Omit<ComponentProps<T>, keyof PropsWithLabels<{}>>
> &
  PropsWithLabels<{}>;
