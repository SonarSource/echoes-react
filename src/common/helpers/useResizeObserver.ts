/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { RefObject, useCallback, useEffect, useState } from 'react';

export function useResizeObserver(ref: RefObject<HTMLElement>) {
  const [width, setWidth] = useState<number>();
  const [height, setHeight] = useState<number>();

  const handleResize = useCallback<ResizeObserverCallback>((entries) => {
    if (!Array.isArray(entries)) {
      return;
    }

    const entry = entries[0];
    setWidth(entry.contentRect.width);
    setHeight(entry.contentRect.height);
  }, []);

  useEffect(() => {
    if (!ref.current) {
      return undefined;
    }

    const RO = new ResizeObserver(handleResize);
    RO.observe(ref.current);

    return () => {
      RO.disconnect();
    };
  }, [handleResize, ref]);

  return { width, height };
}
