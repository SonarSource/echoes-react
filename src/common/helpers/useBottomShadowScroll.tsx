/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import styled from '@emotion/styled';
import { RefObject, useCallback, useEffect, useState } from 'react';
import { cssVar } from '~utils/design-tokens';
import { useResizeObserver } from './useResizeObserver';

/**
 * Custom hook to manage the visibility of a bottom shadow on a scrollable container.
 * The shadow indicates that there is more content to scroll to.
 * @param scrollableContainerRef - Ref to the scrollable container element, its height and scroll event are monitored to recompute the shadow
 * @param resizableContentRef - Ref to the resizable content element inside the scrollable container, its height is monitored to recompute the shadow
 * @returns {boolean} Whether the bottom shadow should be shown
 */
export function useBottomShadowScroll(
  scrollableContainerRef: RefObject<HTMLElement>,
  resizableContentRef: RefObject<HTMLElement>,
) {
  const [showBottomShadow, setShowBottomShadow] = useState(false);

  // Monitor the height of the scrollable container, and recompute the shadow if it changes
  const { height: scrollableContainerHeight } = useResizeObserver(scrollableContainerRef);
  // Monitor the height of the resizable content, and recompute the shadow if it changes, because it may move the scroll position
  const { height: resizableContentHeight } = useResizeObserver(resizableContentRef);

  const updateBottomShadowScroll = useCallback(() => {
    if (scrollableContainerRef.current) {
      const { scrollHeight, clientHeight, scrollTop } = scrollableContainerRef.current;
      setShowBottomShadow(scrollTop + clientHeight < scrollHeight - 8); // -8px for bottom padding
    }
  }, [scrollableContainerRef]);

  // Compute the bottom shadow visibility on first render and attach the scroll event listener
  useEffect(() => {
    const scrollableContainerRefNode = scrollableContainerRef.current;
    if (!scrollableContainerRefNode) {
      return undefined;
    }

    // Re-compute the bottom shadow visibility on scroll event
    scrollableContainerRefNode.addEventListener('scroll', updateBottomShadowScroll);

    // Initial computation
    updateBottomShadowScroll();

    return () => {
      scrollableContainerRefNode.removeEventListener('scroll', updateBottomShadowScroll);
    };
  }, [scrollableContainerRef, updateBottomShadowScroll]);

  // Re-compute the bottom shadow visibility on container and content resize
  useEffect(updateBottomShadowScroll, [
    resizableContentHeight,
    scrollableContainerHeight,
    updateBottomShadowScroll,
  ]);

  return [showBottomShadow] as const;
}

export const BottomShadowScroll = styled.div`
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;

  min-height: ${cssVar('dimension-height-400')};

  border-bottom: ${cssVar('border-width-default')} solid ${cssVar('color-border-weak')};
  background: radial-gradient(
      farthest-side at 50% 100%,
      ${cssVar('bottom-scroll-colors-shadow-gradient')},
      ${cssVar('color-support-transparent')}
    )
    center bottom;
`;
BottomShadowScroll.displayName = 'BottomShadowScroll';
