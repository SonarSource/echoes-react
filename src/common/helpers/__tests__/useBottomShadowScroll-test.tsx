/*
 * Echoes React
 * Copyright (C) 2023-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { act, renderHook, screen } from '@testing-library/react';
import { render } from '../test-utils';
import { BottomShadowScroll, useBottomShadowScroll } from '../useBottomShadowScroll';
import { useResizeObserver } from '../useResizeObserver';

// Mock useResizeObserver since we want to control it in tests
jest.mock('../useResizeObserver', () => ({
  useResizeObserver: jest.fn(),
}));

const SCROLL_HEIGHT = 1000;
const CLIENT_HEIGHT = 400;
const SCROLL_TOP_AT_TOP = 0;
const SCROLL_TOP_NEAR_BOTTOM = SCROLL_HEIGHT - CLIENT_HEIGHT - 20; // A bit before the threshold
const SCROLL_TOP_AT_BOTTOM = SCROLL_HEIGHT - CLIENT_HEIGHT - 4; // Past threshold (should hide shadow)
const CONTAINER_HEIGHT = 500;

// Create a mock HTMLElement
const createMockElement = (
  scrollHeight = SCROLL_HEIGHT,
  clientHeight = CLIENT_HEIGHT,
  scrollTop = SCROLL_TOP_AT_TOP,
) => {
  return {
    scrollHeight,
    clientHeight,
    scrollTop,
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
  } as unknown as HTMLElement;
};

beforeEach(() => {
  jest.clearAllMocks();

  // Reset the mock to return default dimensions
  jest.mocked(useResizeObserver).mockReturnValue({
    width: CONTAINER_HEIGHT,
    height: CONTAINER_HEIGHT,
  });
});

describe('useBottomShadowScroll', () => {
  it('should return initial state of false when not scrollable', () => {
    const scrollableRef = {
      current: createMockElement(CLIENT_HEIGHT, CLIENT_HEIGHT, SCROLL_TOP_AT_TOP),
    };
    const contentRef = { current: createMockElement() };

    const { result } = renderHook(() => useBottomShadowScroll(scrollableRef, contentRef));

    expect(result.current[0]).toBe(false);
  });

  it('should return true when content is scrollable and not at bottom', () => {
    const scrollableRef = {
      current: createMockElement(SCROLL_HEIGHT, CLIENT_HEIGHT, SCROLL_TOP_AT_TOP),
    };
    const contentRef = { current: createMockElement() };

    const { result } = renderHook(() => useBottomShadowScroll(scrollableRef, contentRef));

    expect(result.current[0]).toBe(true);
  });

  it('should hide shadow when scrolled to bottom (within 8px threshold)', () => {
    const scrollableRef = {
      current: createMockElement(SCROLL_HEIGHT, CLIENT_HEIGHT, SCROLL_TOP_AT_BOTTOM),
    };
    const contentRef = { current: createMockElement() };

    const { result } = renderHook(() => useBottomShadowScroll(scrollableRef, contentRef));

    expect(result.current[0]).toBe(false);
  });

  it('should add scroll event listener to scrollable container', () => {
    const mockElement = createMockElement();
    const scrollableRef = { current: mockElement };
    const contentRef = { current: createMockElement() };

    const { unmount } = renderHook(() => useBottomShadowScroll(scrollableRef, contentRef));

    expect(mockElement.addEventListener).toHaveBeenCalledWith('scroll', expect.any(Function));

    unmount();

    expect(mockElement.removeEventListener).toHaveBeenCalledWith('scroll', expect.any(Function));
  });

  it('should update showBottomShadow when scroll event is triggered', () => {
    const mockElement = createMockElement(SCROLL_HEIGHT, CLIENT_HEIGHT, SCROLL_TOP_NEAR_BOTTOM);
    const scrollableRef = { current: mockElement };
    const contentRef = { current: createMockElement() };

    const { result } = renderHook(() => useBottomShadowScroll(scrollableRef, contentRef));

    expect(result.current[0]).toBe(true);

    // Simulate scrolling to bottom
    mockElement.scrollTop = SCROLL_TOP_AT_BOTTOM;

    act(() => {
      // Get the scroll handler that was added and call it
      const scrollHandler = jest.mocked(mockElement.addEventListener).mock
        .calls[0][1] as EventListener;
      scrollHandler(new Event('scroll'));
    });

    expect(result.current[0]).toBe(false);
  });

  it('should update shadow when the containers height changes', () => {
    const scrollableRef = {
      current: createMockElement(CLIENT_HEIGHT, CLIENT_HEIGHT, SCROLL_TOP_AT_TOP),
    };
    const contentRef = { current: createMockElement() };

    const { result, rerender } = renderHook(() => useBottomShadowScroll(scrollableRef, contentRef));

    expect(result.current[0]).toBe(false);

    // Update the scrollableRef to simulate a height change
    scrollableRef.current = createMockElement(SCROLL_HEIGHT, CLIENT_HEIGHT, SCROLL_TOP_AT_TOP);

    // Just a rerender now should trigger a recalculation
    rerender();
    expect(result.current[0]).toBe(false);

    // Mock useResizeObserver to return different container height to trigger the recalculation
    jest.mocked(useResizeObserver).mockReturnValue({
      width: CONTAINER_HEIGHT,
      height: CONTAINER_HEIGHT + 100,
    });
    rerender();

    // Should trigger recalculation
    expect(result.current[0]).toBe(true);
  });
});

describe('BottomShadowScroll component', () => {
  it('should render with correct styling', () => {
    render(<BottomShadowScroll data-testid="bottom-shadow" />);

    expect(screen.getByTestId('bottom-shadow')).toMatchSnapshot();
  });
});
