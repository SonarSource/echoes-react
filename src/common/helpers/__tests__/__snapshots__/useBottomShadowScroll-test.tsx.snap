// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`BottomShadowScroll component should render with correct styling 1`] = `
.emotion-0 {
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  min-height: var(--echoes-dimension-height-400);
  border-bottom: var(--echoes-border-width-default) solid var(--echoes-color-border-weak);
  background: radial-gradient(
        farthest-side at 50% 100%,
        var(--echoes-bottom-scroll-colors-shadow-gradient),
        var(--echoes-color-support-transparent)
      ) center bottom;
}

<div
  class="emotion-0 emotion-1"
  data-testid="bottom-shadow"
/>
`;
